### ISSUE TITLE
HTTP sender scripts not run each time a request is send when using python

### ISSUE THREAD

**Issue by Paippi** [zaproxy/zaproxy#7435]
### Describe the bug

When trying to use python script in HTTP sender using, it only runs on every new request. I'm assuming this is a bug since the script is is run every time when using JavaScript.

### Steps to reproduce the behavior

1. Go to `scripts`
2. From the left pane `scripting` right click on `HTTP Sender`
3. From the spawned option menu select `new script`
4. Write a name for the script, for example, temp.py
5. Select template HttpSender default_template.py
6. Select `Load on Start` and `Enabled`
7. Click `Save`
8. Navigate to a web address using ZAP as a proxy or replay a request using `Request Editor`

### Expected behavior

When using the template default_template.py it should run the script everytime a request is made logging the request and response url everytime a request is send.

### Software versions

OWASP ZAP
Version: 2.11.1
s---
Installed Add-ons: [[id=alertFilters, version=13.0.0],
[id=ascanrules, version=47.0.0], [id=ascanrulesAlpha,
version=38.0.0], [id=ascanrulesBeta, version=41.0.0],
[id=authstats, version=2.0.0], [id=automation,
version=0.16.0], [id=bruteforce, version=11.0.0],
[id=callhome, version=0.4.0], [id=commonlib, version=1.9.0],
[id=communityScripts, version=14.0.0], [id=custompayloads,
version=0.11.0], [id=diff, version=11.0.0],
[id=directorylistv1, version=5.0.0], [id=domxss,
version=13.0.0], [id=encoder, version=0.6.0], [id=exim,
version=0.2.0], [id=fileupload, version=1.1.0],
[id=formhandler, version=5.0.0], [id=fuzz, version=13.6.0],
[id=fuzzdb, version=8.0.0], [id=fuzzdboffensive,
version=4.0.0], [id=gettingStarted, version=13.0.0],
[id=graaljs, version=0.2.0], [id=graphql, version=0.9.0],
[id=help, version=14.0.0], [id=hud, version=0.13.0],
[id=importurls, version=9.0.0], [id=invoke, version=11.0.0],
[id=jruby, version=8.0.0], [id=jwt, version=1.0.2],
[id=jython, version=12.0.0], [id=kotlin, version=1.1.0],
[id=network, version=0.2.0], [id=oast, version=0.10.0],
[id=onlineMenu, version=9.0.0], [id=openapi,
version=27.0.0], [id=pscanrules, version=42.0.0],
[id=pscanrulesAlpha, version=35.0.0], [id=pscanrulesBeta,
version=29.0.0], [id=quickstart, version=33.0.0],
[id=replacer, version=9.0.0], [id=reports, version=0.15.0],
[id=retest, version=0.2.0], [id=retire, version=0.14.0],
[id=reveal, version=4.0.0], [id=saverawmessage,
version=7.0.0], [id=savexmlmessage, version=0.3.0],
[id=scripts, version=30.0.0], [id=selenium, version=15.9.0],
[id=soap, version=13.0.0], [id=spiderAjax, version=23.8.0],
[id=tips, version=9.0.0], [id=tokengen, version=15.0.0],
[id=wappalyzer, version=21.12.0], [id=webdriverlinux,
version=42.0.0], [id=websocket, version=26.0.0], [id=zest,
version=35.0.0]]
---
Operating System: Linux
Java Version: Ubuntu 11.0.16
System's Locale: en_US
Display Locale: en_GB
Format Locale: en_US
ZAP Home Directory: /home/<redacted>/.ZAP/
ZAP Installation Directory: /opt/zaproxy/./
Look and Feel: FlatLaf Light (com.formdev.flatlaf.FlatLightLaf)

### Would you like to help fix this issue?

- [ ] Yes

**Comment by Paippi** [comment]
Sure, it's the template provided. Here's the python script I used:
```python
# The sendingRequest and responseReceived functions will be called for all
# requests/responses sent/received by ZAP, including automated tools (e.g.
# active scanner, fuzzer, ...)

# Note that new HttpSender scripts will initially be disabled
# Right click the script in the Scripts tree and select "enable"

# 'initiator' is the component the initiated the request:
#      1   PROXY_INITIATOR
#      2   ACTIVE_SCANNER_INITIATOR
#      3   SPIDER_INITIATOR
#      4   FUZZER_INITIATOR
#      5   AUTHENTICATION_INITIATOR
#      6   MANUAL_REQUEST_INITIATOR
#      7   CHECK_FOR_UPDATES_INITIATOR
#      8   BEAN_SHELL_INITIATOR
#      9   ACCESS_CONTROL_SCANNER_INITIATOR
#     10   AJAX_SPIDER_INITIATOR
# For the latest list of values see the HttpSender class:
# https://github.com/zaproxy/zaproxy/blob/main/zap/src/main/java/org/parosproxy/paros/network/HttpSender.java
# 'helper' just has one method at the moment: helper.getHttpSender() which
# returns the HttpSender instance used to send the request.
#
# New requests can be made like this:
# msg2 = msg.cloneAll() # msg2 can then be safely changed without affecting msg
# helper.getHttpSender().sendAndReceive(msg2, false)
# print('msg2 response code =' + msg2.getResponseHeader().getStatusCode())


def sendingRequest(msg, initiator, helper):
    # Debugging can be done using print like this
    print('sendingRequest called for url=' +
          msg.getRequestHeader().getURI().toString())


def responseReceived(msg, initiator, helper):
    # Debugging can be done using print like this
    print('responseReceived called for url=' +
          msg.getRequestHeader().getURI().toString())
```
And here's the JavaScript code I used to compare:
```javascript
// The sendingRequest and responseReceived functions will be called for all requests/responses sent/received by ZAP, 
// including automated tools (e.g. active scanner, fuzzer, ...)

// Note that new HttpSender scripts will initially be disabled
// Right click the script in the Scripts tree and select "enable"  

// 'initiator' is the component the initiated the request:
// 		1	PROXY_INITIATOR
// 		2	ACTIVE_SCANNER_INITIATOR
// 		3	SPIDER_INITIATOR
// 		4	FUZZER_INITIATOR
// 		5	AUTHENTICATION_INITIATOR
// 		6	MANUAL_REQUEST_INITIATOR
// 		7	CHECK_FOR_UPDATES_INITIATOR
// 		8	BEAN_SHELL_INITIATOR
// 		9	ACCESS_CONTROL_SCANNER_INITIATOR
// 		10	AJAX_SPIDER_INITIATOR
// For the latest list of values see the HttpSender class:
// https://github.com/zaproxy/zaproxy/blob/main/zap/src/main/java/org/parosproxy/paros/network/HttpSender.java
// 'helper' just has one method at the moment: helper.getHttpSender() which returns the HttpSender 
// instance used to send the request.
//
// New requests can be made like this:
// msg2 = msg.cloneAll() // msg2 can then be safely changed as required without affecting msg
// helper.getHttpSender().sendAndReceive(msg2, false);
// print('msg2 response=' + msg2.getResponseHeader().getStatusCode())

function sendingRequest(msg, initiator, helper) {
	// Debugging can be done using println like this
	print('sendingRequest called for url=' + msg.getRequestHeader().getURI().toString())
}

function responseReceived(msg, initiator, helper) {
	// Debugging can be done using println like this
	print('responseReceived called for url=' + msg.getRequestHeader().getURI().toString())
}
```

**Comment by kingthorin** [comment]
Thanks. Also the requests are definitely going through ZAP in both cases. They aren't missing because they're being cached by the browser and not actually going through ZAP or something like that?

**Comment by Paippi** [comment]
Yes, I'm pretty sure that's the case since comparing to JavaScript it does work differently and I've tested with both the request editor and a browser (Firefox) using force refresh.

**Comment by thc202** [comment]
How are you telling if it's being executed or not?

**Comment by Paippi** [comment]
Comparing the logs the script prints in the `Script Console`

**Comment by thc202** [comment]
Issue title updated to better reflect the issue, the scripts are being executed it's not printing to the Script Console but ZAP's output on following runs.

**Comment by thc202** [comment]
For the record this was raised in the Jython issue tracker: jython/jython#209.
